cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+Comparability revolves around the inequality operators (`>`, `<`, `>=`, `<=`), and defines the semantics they need to properly determine how two values compare to each other.
+Orderability revolves around the `ORDER BY` clause, and defines the semantics it needs to be able to properly sort values passed to it.
+Futhermore, the definitions of *equality* and *equivalence* are harmonized and spelled out.
+In summary, we are proposing some changes to how Cypher defines these four concepts in order to get a consistent set of rules and a simpler conceptual model.
+--
+
+toc::[]
+
+
+== Motivation
+
+There are currently a number of limitations and inconsistencies that this CIP aims to fix.
+
+(1) Cypher already has good semantics for equality within the primitive types (booleans, strings, integers, and floats) and maps.
+Furthermore, Cypher has good semantics for comparability and orderability for integers, floats, and strings, within the types.
+However working with values of different types can be difficult:
+
+  * Comparability between values of different types is often undefined, which is something we regard as a problem for certain combinations of values and types.
+  * `ORDER BY` will often fail with an error if the values passed to it have different types.
+
+(2) The underlying conceptual model is complex and sometimes inconsistent which leads to an unclear relationship between comparison operators, equality, grouping, and `ORDER BY`:
+
+  * Comparability and orderability are not aligned with each other consistently (Some types may be ordered but not compared).
+  * There are various inconsistencies around equality (and equivalence) semantics as exposed by `IN`, `=`, `DISTINCT`, and grouping.
+  The difference between equality and equivalence in Cypher today is small and subtle, and limited to testing two instances of the value `null` to each other.
+
+    ** In equality, `null = null` is `null`.
+    ** In equivalence, used by grouping and `DISTINCT`, instances of `null` are always treated as being the same value.
+    ** However, equality treats `null` values differently if they are an element of a list or a map value.
+    ** Similar rules apply for `NaN`.
+
+== Proposal
+
+=== Comparability
+
+We propose that comparability be defined between any pair of values, as specified below (Please note that `null` is interpreted as meaning "unknown" or "could be any value" in the context of comparability):
+
+* Integers are compared numerically.
+* Floats (excluding `NaN` and the Infinities) are compared numerically.
+* Arbitrary numbers (excluding `NaN` and the Infinities) are compared to each other numerically, i.e. the comparison behaves as if both numbers would have been coerced to arbitrary precision big decimals (currently outside the Cypher type system) before comparing them with each other numerically.
```

Comment:
Given that integers and floats have been mentioned already, what precisely is an "arbitrary" number?